zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/LearnerHandler.java

```diff
@@ -820,11 +820,25 @@ boolean syncFollower(long peerLastZxid, LearnerMaster learnerMaster) {
                     currentZxid = queueCommittedProposals(txnLogItr, peerLastZxid,
                                                          minCommittedLog, maxCommittedLog);
 
-                    LOG.debug("Queueing committedLog 0x" + Long.toHexString(currentZxid));
-                    Iterator<Proposal> committedLogItr = db.getCommittedLog().iterator();
-                    currentZxid = queueCommittedProposals(committedLogItr, currentZxid,
-                                                         null, maxCommittedLog);
-                    needSnap = false;
+                    if (currentZxid < minCommittedLog) {
+                        LOG.info("Detected gap between end of txnlog: " +
```

Comment:
I am curious to know if we have any real world data about how often this happens and if this happens, how many cases leader would able to recover (and how long it would take). I would imagine if majority of cases a leader fail to catch up, then it would be optimal for us to force a leader election, rather than keep sending snapshots for a sync (especially if snapshot is big). Any thought on this?

Reply:
On the other side, since this clearly fix a correctness issue in current code, I am +1 on land this if we don't have better trade off to make here.

Reply:
Unfortunately, we do not have data on how often this problem occurs. A (small) spot check on our production logs didn't turn up any instances but, if memory serves, this was an important protection in the extreme case. A leader does not perpetually hit this gap code so deposing it strikes me as overkill.